lib/mayaUsd/ufe/private/Utils.cpp

```diff
@@ -129,54 +129,49 @@ UsdGeomXformCommonAPI convertToCompatibleCommonAPI(const UsdPrim& prim)
 
 void applyCommandRestriction(const UsdPrim& prim, const std::string& commandName)
 {
-    // early check to see if a particular node has any specs to contribute
-    // to the final composed prim. e.g (a node in payload)
-    if(!MayaUsdUtils::hasSpecs(prim)){
+    auto primSpec = MayaUsdUtils::getPrimSpecAtEditTarget(prim);
+    auto primStack = prim.GetPrimStack();
+    std::string layerDisplayName;
 
-        auto layers = MayaUsdUtils::layerInCompositionArcsWithSpec(prim);
-        std::string layerDisplayNames;
-        for (auto layer : layers) {
-            layerDisplayNames.append("[" + layer->GetDisplayName() + "]" + ",");
-        }
-        layerDisplayNames.pop_back();
-        std::string err = TfStringPrintf("Cannot %s [%s]. It does not make any contributions in the current layer "
-                                         "because its specs are in an external composition arc. Please open %s to make direct edits.",
-                                         commandName.c_str(),
-                                         prim.GetName().GetString().c_str(), 
-                                         layerDisplayNames.c_str());
-        throw std::runtime_error(err.c_str());
-    }
-
-    // if the current layer doesn't have any contributions
-    if (!MayaUsdUtils::doesEditTargetLayerContribute(prim)) {
-        auto strongestContributingLayer = MayaUsdUtils::strongestContributingLayer(prim);
-        std::string err = TfStringPrintf("Cannot %s [%s]. It is defined on another layer. Please set [%s] as the target layer to proceed.", 
-                                         commandName.c_str(),
-                                         prim.GetName().GetString().c_str(),
-                                         strongestContributingLayer->GetDisplayName().c_str());
-        throw std::runtime_error(err.c_str());
-    }
-    else
+    // iterate over the prim stack, starting at the highest-priority layer.
+    for (const auto& spec : primStack) 
     {
-        auto layers = MayaUsdUtils::layersWithContribution(prim);
-        // if we have more than 2 layers that contributes to the final composed prim
-        if (layers.size() > 1) {
-            std::string layerDisplayNames;
+        const auto& layerName = spec->GetLayer()->GetDisplayName();
 
-            // skip the the first arc which is PcpArcTypeRoot
-            // we are interested in all the arcs after root
-            std::for_each(std::next(layers.begin()),layers.end(), [&](const auto& it) {
-                 layerDisplayNames.append("[" + it->GetDisplayName() + "]" + ",");
-            });
+        if (primSpec != spec) {
+            layerDisplayName.append("[" + layerName + "]" + ",");
+        }
+        else {
+            // if exist a primSpec with reference
+            if(spec->HasReferences()) {
```

Comment:
Should comment on why we abort the loop here.

Reply:
The main reason for the skipping the reference is to not clash with the **over** that may be created in **the stage's local sessioLayer or other layers**.

Consider the case, where we have a stage that has opened a USD file called **RefExample.usda** and in that file we have 
a prim with **over** which has an external reference to another Usd file called **HelloWorld.usda**.

**RefExample.usda**
```
over "refSphere2" (
    prepend references = @./HelloWorld.usda@
)
```

**HelloWorld.usda**
```
#usda 1.0
(
    defaultPrim = "hello"
)

def Xform "hello"
{
	...........
}
```

**All PrimStack knows at this point is that "refSphere2" has two specs in 2 different places in order from strong to weak:**

```
RefExample.usda  --->  over
HelloWorld.usda  --->  def
```

Now in line 161, I unconditionally check if there is an override and then record it's layer name which then results in error message to be displayed which is not desirable:

```
// if exists an over sepc
if (spec->GetSpecifier() == SdfSpecifierOver) {
    layerDisplayName.append("[" + layerName + "]" + ",");
}
```

one should be able to rename "refSphere2" since it is created in "RefExample.usda" and itself doesn't have any other "over" in the stage's local layers.

This is why I previously had a condition to check if the "over" was created in sessioLayer or not.

Hope this makes sense.

Reply:
It does.  Can you distill this into a code comment somehow?  I know it's a bit challenging, you don't want to have a 2-page code comment, but it would still be nice.

Reply:
Ugh yes of course! I was planning to do it after the planning meeting.